server-mux.go

```diff
@@ -0,0 +1,214 @@
+			c.dataBuf.unRead = false
+			c.dataBuf.offset = 0
+			c.dataBuf.buffer = c.dataBuf.buffer[:]
+			if n < len(b) {
+				// Continue copying from socket if b still has room for data
+				tmpBuffer := make([]byte, len(b)-n-1)
+				nr, err := c.Conn.Read(tmpBuffer)
+				for idx, val := range tmpBuffer {
+					b[n+idx] = val
+				}
+				return n + nr, err
+			}
+		}
+		// We here return the last error
+		return n, c.lastError
+	}
+	return c.Conn.Read(b)
+}
+
+// MuxListener - encapuslates the standard net.Listener to inspect
+// the communication protocol upon network connection
+type MuxListener struct {
+	net.Listener
+	config *tls.Config
+}
+
+// NewMuxListener - creates new MuxListener, returns error when cert/key files are not found
+// or invalid
+func NewMuxListener(listener net.Listener, certPath, keyPath string) (MuxListener, error) {
+	var err error
+	config := &tls.Config{}
+	if config.NextProtos == nil {
+		config.NextProtos = []string{"http/1.1", "h2"}
+	}
+	config.Certificates = make([]tls.Certificate, 1)
+	config.Certificates[0], err = tls.LoadX509KeyPair(certPath, keyPath)
+	if err != nil {
+		return MuxListener{}, err
+	}
+	return MuxListener{Listener: listener, config: config}, nil
+}
+
+// Accept - peek the protocol to decide if we should wrap the
+// network stream with the TLS server
+func (l MuxListener) Accept() (net.Conn, error) {
+	c, err := l.Listener.Accept()
+	if err != nil {
+		return c, err
+	}
+	cmux := NewMuxConn(c)
+	protocol := cmux.PeekProtocol()
+	if protocol == "tls" {
+		return tls.Server(cmux, l.config), nil
+	}
+	return cmux, nil
+}
+
+// MuxServer - the main mux server
+type MuxServer struct {
+	http.Server
+}
+
+// ListenAndServeTLS - similar to the http.Server version. However, it has the
+// ability to redirect http requests to the correct HTTPS url if the client
+// mistakenly initiates a http connection over the https port
+func (m *MuxServer) ListenAndServeTLS(certFile, keyFile string) error {
+	listener, err := net.Listen("tcp", m.Server.Addr)
+	if err != nil {
+		return err
+	}
+	mux, err := NewMuxListener(listener, mustGetCertFile(), mustGetKeyFile())
+	if err != nil {
+		return err
+	}
+	err = http.Serve(mux,
+		http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			// We reach here when MuxListener.MuxConn is not wrapped with tls.Server
+			if r.TLS == nil {
+				u := url.URL{
+					Scheme:   "https",
+					Opaque:   r.URL.Opaque,
+					User:     r.URL.User,
+					Host:     r.Host,
+					Path:     r.URL.Path,
+					RawQuery: r.URL.RawQuery,
+					Fragment: r.URL.Fragment,
+				}
+				http.Redirect(w, r, u.String(), http.StatusMovedPermanently)
+			} else {
+				// Execute registered handlers
+				m.Server.Handler.ServeHTTP(w, r)
+			}
+		}))
+	return err
+}
+
+// ListenAndServe - Same as the http.Server version
+func (m *MuxServer) ListenAndServe() error {
+	return m.Server.ListenAndServe()
+}
+
+func longestWord(strings []string) int {
```

Comment:
> Just out of curiosity: Instead of cooking up own implementation for multiplexing, couldn't we use an alternative (such as https://github.com/soheilhy/cmux).

No it was on purpose to not use it the cmux API's are not convenient and we don't need a complex implementation like what they have.